### ISSUE TITLE
How to ensure in source plugin that subsequent gatsby builds result in updated pages?

### ISSUE THREAD

**Issue by datakurre** [gatsbyjs/gatsby#6756]
I'm trying to optimize our Gatsby source plugin to properly update data on subsequent builds so that only changed data is fetched from the backend (setPluginStatus to set lastFetched timestamp and then touchNode to touch unchanged nodes and createNode to add and update nodes).

Currently I have an issue where updating node (with createNode) does not always result in updated data in created pages.

I'm on `"gatsby": "^2.0.0-beta.59"`.

I create pages programmatically similarly to https://next.gatsbyjs.org/tutorial/part-seven/#part-seven

I can `grep` that outdated data cannot be found in `./.cache`, but only in `./public`.

Updating some React component does update page with that component, but page is still being rendered with outdated data.

Running clean build results in correct data.

Any tips, how I could debug this further? Or what could be the cause?

**Comment by pieh** [comment]
Gatsby is looking into `node.internal.contentDigest` and compare it with value of that field in previously stored node. If this field didn't change that signals to Gatsby that node didn't change and it won't rerun queries for pages that are using it and will reuse results from previous builds.

Common way to generate it is using `crypto` module md5 hash:
```js
const nodeContentDigest = crypto
      .createHash('md5')
      .update(JSON.stringify(nodeData))
      .digest('hex')
```
or if your backend have feature like that baked in, you could use that.

If this doesn't work, please create bug report with minimal reproduction example

**Comment by datakurre** [comment]
@pieh Thanks.

**Comment by datakurre** [comment]
@pieh I can confirm that `./cache/redux-state.json` gets updated with a new node contentDigest, but related page build from that node is not updated.

Is there a way for simulate update just by mutating redux-state manually? (assuming the source plugin would just touch all nodes)

I now assume that this is a bug, and try later to reproduce it with gatsby starter.

**Comment by datakurre** [comment]
@pieh Thanks. This was not a bug after all, but merely improper use of API. I believe.

I've been figuring out, how to update a node in the middle of content tree. After the first update with  `createNode` I was greeted by an error about (touched) children no longer being found during the build.

I could not find proper documentation for that, so I just played around and ended up calling `createParentChildLink` (before `createNode`) to avoid the children from being invalidated (build stopped failing, so I did not see my error).

Now I found out that, because of calling `createParentChildLink` before calling `createNode`, `contentDigest` was already updated before calling `createNode` and the node got flagged as not changed.

That's why I ended up having the new `contentDigest` in store, but outdated pages.
